[BIOMAGE-1677] Make aggregated pseudobulk matrix

gerbeldo opened the pull request:
# Description
Implements a function to aggregate count matrices, to use in pseudobulk differential expression.

# Details
#### URL to issue
https://biomage.atlassian.net/browse/BIOMAGE-1677

#### Link to staging deployment URL (or set N/A)
N/A

#### Links to any PRs or resources related to this PR
<!---
  Delete this comment and include the URLs of any pull requests that are related to this PR.
  Place each PR on a new line.
-->

#### Integration test branch
master
<!---
  The branch of the integration test this PR will be run against

  If you DID NOT modify the integration tests for this PR, this can be left as `master`.

  If you DID modify the integration tests for this PR, add the name of the branch you created
  in hms-dbmi-cellenics/testing that will be used to test this branch.
-->

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.
<!---
  The required checks will not pass until all the boxes below have been checked.
-->

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [ ] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [x] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [x] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [x] Unit tests written **or** no unit tests required for change, e.g. documentation update.

<!---
  Download the latest production data using `biomage experiment pull`.
  To set up easy local testing with inframock, follow the instructions here: https://github.com/hms-dbmi-cellenics/inframock
  To deploy to the staging environment, follow the instructions here: https://github.com/hms-dbmi-cellenics/biomage-utils
-->

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [x] Started end-to-end tests on the latest commit.

### Documentation updates
- [x] Relevant Github READMEs updated **or** no GitHub README updates required.
- [x] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [X] Photo of a cute animal attached to this PR.
<img width="359" alt="Screen Shot 2022-02-01 at 14 22 07" src="https://user-images.githubusercontent.com/21223700/152018208-3751c202-6d53-431b-8af7-0b3a665c3153.png">

gerbeldo committed:
initial seurat implementation

gerbeldo committed:
reimplement, using presto for aggregation

gerbeldo committed:
change sample slots to `samples`

gerbeldo committed:
add function description

gerbeldo committed:
tidy function calls. specify packages.

gerbeldo commented on r/R/make_pseudobulk_matrix.R:
```diff
@@ -0,0 +1,41 @@
+
+  # create groups for aggregation
+  pbulk_groups <-
+    factor(paste(scdata$samples, scdata$custom, sep = "_"))
```
This should use a better separator. samples usually contain underscores.

alexvpickering replied:
once you subset scdata you can just do the aggregation on samples

gerbeldo replied:
yeah! that was one of the things I wasn't sure about. that makes it simpler. I'll change it

gerbeldo commented on r/R/make_pseudobulk_matrix.R:
```diff
@@ -0,0 +1,41 @@
+
+  # recover metadata
+  pbulk_metadata <- colnames(agg)
+  pbulk_metadata <- data.table::tstrsplit(pbulk_metadata, split = "_")
```
same here when splitting

gerbeldo committed:
add return to function

gerbeldo committed:
filter cells NAs in custom slot

gerbeldo committed:
amends documentation

gerbeldo committed:
adds a space

gerbeldo committed:
removes sample/custom label generation.

After filtering cells with NA, we can group by sample. To recover
original metadata we can join to a group_key table.

gerbeldo committed:
add seurat call

alexvpickering commented on r/R/make_pseudobulk_matrix.R:
```diff
@@ -0,0 +1,62 @@
+#'
+makePseudobulkMatrix <- function(scdata) {
+  # filter out cells not in base/background groups
+  scdata <- scdata[, !is.na(scdata@meta.data$custom)]
```
```suggestion
  scdata <- scdata[, !is.na(scdata$custom)]
```

alexvpickering commented on r/R/make_pseudobulk_matrix.R:
```diff
@@ -0,0 +1,62 @@
+  counts <- scdata[["RNA"]]@counts
+  gene_annotations <- scdata@misc$gene_annotations
+
+  groups <- factor(scdata@meta.data$samples)
```
```suggestion
  groups <- factor(scdata$samples)
```

alexvpickering commented on r/R/make_pseudobulk_matrix.R:
```diff
@@ -0,0 +1,62 @@
+  rownames(metadata) <- NULL
+
+  return(metadata)
+}
```
```suggestion
  # aggregate over samples
  samples <- scdata$samples
  samples <- factor(samples, levels = unique(samples))

  agg <- presto::sumGroups(counts, samples, MARGIN = 1)
  agg <- Matrix::Matrix(agg, sparse = TRUE)
  agg <- Matrix::t(agg)

  #row/colnames are lost in aggregation
  rownames(agg) <- rownames(counts)
  colnames(agg) <- levels(samples)

  # recover original metadata
  metadata <- data.frame(samples = colnames(agg),
                         custom = scdata$custom[!duplicated(samples)],
                         row.names = colnames(agg))

  # create seurat, and add metadata
  pbulk <- Seurat::CreateSeuratObject(agg, meta.data = metadata)
  pbulk@misc$gene_annotations <- gene_annotations

  return(pbulk)
```

A couple of comments:
- have to be careful with letting levels be set by `factor` as they are by default sorted into increasing order. I think this is why you had the inner join and key lookup by you can probably avoid by just explicitly setting the levels to the order that items appear in
- other than that looks good, I just cleaned it up a bit

gerbeldo committed:
Update r/R/make_pseudobulk_matrix.R

Co-authored-by: Alex Pickering <alexvpickering@gmail.com>

gerbeldo committed:
remove @meta.data from calls to metadata

Co-authored-by: Alex Pickering <alexvpickering@gmail.com>

gerbeldo committed:
careful level spec and remove group key function

Co-authored-by: Alex Pickering <alexvpickering@gmail.com>

gerbeldo committed:
added missing bracket, lint

gerbeldo committed:
add mock data, some tests

gerbeldo committed:
styler

gerbeldo committed:
add correct aggregation test

gerbeldo committed:
styler

gerbeldo committed:
added package to fun calls

alexvpickering commented on r/tests/testthat/test-make-pseudobulk-matrix.R:
```diff
@@ -0,0 +1,86 @@
+  # add cells not belonging to groups to compare
+  scdata$custom[floor(ncol(scdata) / 10) * 1:10] <- NA
+
+  scdata
```
```suggestion
  return(scdata)
```

alexvpickering commented on r/tests/testthat/test-make-pseudobulk-matrix.R:
```diff
@@ -0,0 +1,86 @@
+
+  expected_rows <- nrow(scdata)
+  expected_cols <-
+    length(unique(scdata@meta.data$samples[!is.na(scdata@meta.data$custom)]))
```
```suggestion
    length(unique(scdata$samples[!is.na(scdata$custom)]))
```

alexvpickering commented on r/tests/testthat/test-make-pseudobulk-matrix.R:
```diff
@@ -0,0 +1,86 @@
+mock_scdata <- function() {
+  scdata <- pbmc_small
```
```suggestion
  data("pbmc_small", package = "SeuratObject", envir = environment())
  scdata <- pbmc_small
```

alexvpickering reviewed (approved):
LGTM. see suggestions, nice work!

gerbeldo committed:
Update r/tests/testthat/test-make-pseudobulk-matrix.R

Co-authored-by: Alex Pickering <alexvpickering@gmail.com>

gerbeldo committed:
Update r/tests/testthat/test-make-pseudobulk-matrix.R

Co-authored-by: Alex Pickering <alexvpickering@gmail.com>

gerbeldo committed:
Update r/tests/testthat/test-make-pseudobulk-matrix.R

Co-authored-by: Alex Pickering <alexvpickering@gmail.com>
